### PR TITLE
OCPBUGS-861: Rebase openshift/etcd 4.12 onto v3.5.5

### PR DESCRIPTION
**Pull request by tjungblu**
This was rebased with:

> git rebase --rebase-merges --fork-point v3.5.3 v3.5.5

this seems to effectively pick all commits between the tags and apply them onto the branch HEAD including the merge commits. thanks @hasbro17

### CODE REVIEW

**Comment by tjungblu** [comment]
/retitle Rebase onto 3.5.4

keeping the bug out of this story here, as there is no 4.13 fix version in jira available. We need the bug ticket for the 4.11 PR though

**Comment by Elbehery** [comment]
nice work @tjungblu  

Can we describe the steps taken, and how to verify its correctness ?

**Comment by tjungblu** [comment]
@Elbehery please read https://github.com/openshift/etcd/pull/141

**Comment by tjungblu** [comment]
/retitle OCPBUGS-861: Rebase openshift/etcd 4.12 onto v3.4.20

**Comment by tjungblu** [comment]
/retitle OCPBUGS-861: Rebase openshift/etcd 4.12 onto v3.5.4

**Comment by hasbro17** [comment]
@tjungblu Looking at the diff between 3.5.4 and 3.5.3 we have some merge commits as well.
https://github.com/etcd-io/etcd/compare/v3.5.3...v3.5.4

Does rebasing the 3.5.4 tag ontop of openshift-4.12 drop those merge commits, or is there a way to preserve them.
Seems like we did have them preserved in prior rebases/merges.
https://github.com/openshift/etcd/commits/openshift-4.12?after=31b6b2d9b4d7351088d74d80362316f515cba037+34&branch=openshift-4.12&qualified_name=refs%2Fheads%2Fopenshift-4.12

**Comment by hasbro17** [comment]
Maybe [`--rebase-merges`](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---rebase-mergesrebase-cousinsno-rebase-cousins) to preserve the merge commits as well?

**Comment by tjungblu** [comment]
I (believe) this went through without a merge commit, as there was no conflict on the files since the last merge. Imagine you just cherry pick those four commits in sequence without conflicts. 

Or do you mean that the merge commits from previous rebases got lost?

**Comment by hasbro17** [comment]
I meant as we were cherry picking the commits between 3.5.3 and 3.5.4 we currently exclude the "merge commits" that are generated from when the upstream PRs merged. I think when you run a rebase command without `--rebase-merges` it drops those commits by default. In order to preserve the full history I think we should include those.

As an example this is what adding the `git rebase --rebase-merges --fork-point v3.5.3 v3.5.4` yields on the `openshift-4.12` branch:
https://github.com/openshift/etcd/compare/openshift-4.12...hasbro17:etcd:rebase-3.5.4-on-top-of-4.12?expand=1

This way we keep the "merge pull request commits" from upstream like we have previously for all the release branches.
E.g: https://github.com/openshift/etcd/commit/12ceac6fdd8e019274c822aea3ef3488d180ad56

**Comment by tjungblu** [comment]
got it, I'll send you a new one for the v3.5.5 with the merges included. Thanks for looking into that @hasbro17

**Comment by tjungblu** [comment]
@hasbro17 as with the 4.8 rebase, seems we need to bump the golang version as well:
https://github.com/openshift/release/pull/32217

**Comment by tjungblu** [comment]
/hold

waiting for https://github.com/openshift/release/pull/32217 and we need to coordinate with the k8s rebase.

**Comment by tjungblu** [comment]
K8s is merged, we can continue with the payload testing here. 

https://github.com/openshift/release/pull/32217 should merge soonish hopefully

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -44,6 +44,7 @@ jobs:
             GOARCH=amd64 PASSES='build' ./test.sh
             GOARCH=386 PASSES='build' ./test.sh
             GO_BUILD_FLAGS='-v -mod=readonly' GOOS=darwin GOARCH=amd64 ./build.sh
+            GO_BUILD_FLAGS='-v -mod=readonly' GOOS=darwin GOARCH=arm64 ./build.sh
             GO_BUILD_FLAGS='-v -mod=readonly' GOOS=windows GOARCH=amd64 ./build.sh
```
Comment (EmilyM1) [inline .github/workflows/tests.yaml:48]: Hi -- are there 2 flags for darwin amd64?

Comment (hasbro17) [reply .github/workflows/tests.yaml:48]: That second arch flag for darwin is `arm64` not `amd64` i.e Apple Silicon or M1 cpus:
https://tip.golang.org/doc/go1.16#ports
https://github.com/etcd-io/etcd/pull/14436

**Comment by hasbro17** [comment]
/approve
/payload 4.12 ci blocking

Informing jobs look good. The upgrade blocking ones seems unrelated?:
```
[36mINFO[0m[2022-10-04T13:05:50Z] Running step release-analysis-prpqr-aggregator-openshift-release-analysis-prpqr-aggregator. 
[36mINFO[0m[2022-10-04T14:29:41Z] Step release-analysis-prpqr-aggregator-openshift-release-analysis-prpqr-aggregator failed after 1h23m50s.
```

@tjungblu I think 4.12 and 4.13 are the same but I'm thinking we probably need to go through 4.13 first?
>expected [Jira Issue OCPBUGS-861](https://issues.redhat.com//browse/OCPBUGS-861) to depend on a bug targeting a version in 4.13.0 and in one of the following states: MODIFIED, ON_QA, VERIFIED, but no dependents were found

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (tjungblu) [reply .github/workflows/tests.yaml:48]: yep etcd 3.5.5 has m1 support now

**Comment by tjungblu** [comment]
> @tjungblu I think 4.12 and 4.13 are the same but I'm thinking we probably need to go through 4.13 first?

@hasbro17 I've raised this before with Rory, apparently Sam was always just overriding the BZ check. I've put this into https://issues.redhat.com/browse/ETCD-316

**Comment by tjungblu** [comment]
looking into the failure in 
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/openshift-etcd-144-nightly-4.12-e2e-aws-sdn-serial/1577619078351163392

**Comment by tjungblu** [comment]
seems there was some networking trouble between peers for some time: 

image

the errors are fine:

image

I'm a bit afraid this health event is too spammy in CEO, I've seen something similar here:
https://issues.redhat.com/browse/OCPBUGS-1128?focusedCommentId=20991825&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20991825

The events are very spaced out though, they were at 11:44, then 11:49/12:01 then 12:58, 13:28. Can't tell from the etcd log what made it return "unhealthy" however.

**Comment by tjungblu** [comment]
okay there was a timeout when creating the client:

```
2022-10-05T11:44:06.460738407Z E1005 11:44:06.460693       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-176-100.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.176.100:2379]: context deadline exceeded
2022-10-05T11:47:06.466633212Z E1005 11:47:06.466579       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-252-170.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.252.170:2379]: context deadline exceeded
2022-10-05T11:49:06.516364638Z E1005 11:49:06.516318       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-176-100.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.176.100:2379]: context deadline exceeded
2022-10-05T11:51:06.499940889Z E1005 11:51:06.499895       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-137-36.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.137.36:2379]: context deadline exceeded
2022-10-05T11:59:06.453666297Z E1005 11:59:06.453612       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-252-170.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.252.170:2379]: context deadline exceeded
2022-10-05T12:01:06.467145573Z E1005 12:01:06.467091       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-176-100.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.176.100:2379]: context deadline exceeded
2022-10-05T12:03:06.479799318Z E1005 12:03:06.479754       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-137-36.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.137.36:2379]: context deadline exceeded
2022-10-05T12:56:06.501553616Z E1005 12:56:06.501506       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-252-170.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.252.170:2379]: context deadline exceeded
2022-10-05T12:58:06.513221056Z E1005 12:58:06.513174       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-176-100.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.176.100:2379]: context deadline exceeded
2022-10-05T13:00:06.537830871Z E1005 13:00:06.537773       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-137-36.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.137.36:2379]: context deadline exceeded
2022-10-05T13:02:06.746612899Z E1005 13:02:06.746559       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-151-155.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.151.155:2379]: context deadline exceeded
2022-10-05T13:26:06.512965255Z E1005 13:26:06.512905       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-252-170.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.252.170:2379]: context deadline exceeded
2022-10-05T13:28:06.525243345Z E1005 13:28:06.525181       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-176-100.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.176.100:2379]: context deadline exceeded
2022-10-05T13:30:06.528854040Z E1005 13:30:06.528796       1 etcdmemberscontroller.go:73] Unhealthy etcd member found: ip-10-0-137-36.ec2.internal, took=, err=create client failure: failed to make etcd client for endpoints [https://10.0.137.36:2379]: context deadline exceeded

```

EDIT: this correlates with the revision rollouts, so nothing problematic. We need to adjust the CEO behaviour here though...

**Comment by hasbro17** [comment]
/hold
@tjungblu feel free to unhold once we're clear on the nightly payload ci run.
Although we're already just waiting for aws-serial to clear which is the same.

**Comment by Elbehery** [comment]
> seems there was some networking trouble between peers for some time:
> 
> image
> 
> the errors are fine:
> 
> image
> 
> I'm a bit afraid this health event is too spammy in CEO, I've seen something similar here: https://issues.redhat.com/browse/OCPBUGS-1128?focusedCommentId=20991825&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20991825
> 
> The events are very spaced out though, they were at 11:44, then 11:49/12:01 then 12:58, 13:28. Can't tell from the etcd log what made it return "unhealthy" however.

how did u get this prometheus ? i am just curious to use the same approach debugging my failed ci :)

**Comment by tjungblu** [comment]
@Elbehery https://promecieus.dptools.openshift.org/?search= just paste the prow url

**Comment by tjungblu** [comment]
guess everybody is merging on the last day of the release today. I'll retry, maybe it goes through over the night :)

/hold cancel

**Pull request merged**